### ISSUE TITLE
Performance of `forall` vs `for` in Domain Expansion

### ISSUE THREAD

**Issue by Tshimanga** [chapel-lang/chapel#8542]
### Summary of Problem
In trying to use `forall` to expand a domain from iterations over a large set, `forall` at best runs twice as slow as `for`. The lower performance persists through 100,000,000 iterations. Is there a way to benefit from parallelization when constructing domains and arrays from large sets? I pushed up a repo at https://github.com/Tshimanga/chapel-loops/tree/develop which contains executables  #compiled from the code below. `nforall` uses `with (+ reduce indices)`, `allforn` uses `with (ref indices)`, and `nloopsinvacuum` runs empty loops for reference.

At n=10,000 (10^8 iterations) performance was as follows:
Serial ~ 130s
Parallel w/ `reduce` ~ 580s
Parallel w/ `ref` ~ 230s

At n=100,000 the process dies.

These results come from running the processes on 8 cores.

### Steps to Reproduce

The executables are derived from the code below whose purpose was to test `forall`'s performance with respect to `for` in the task of building domains/arrays.

```chapel
use Time;

config const n = 1000;

var indices1: domain(2*int);
var indices2: domain(2*int);

var t1: Timer;
var t2: Timer;

var D = {1..n, 1..n};
t1.start();
for (i,j) in D {
    indices1 += (i,j);
}
t1.stop();


t2.start();
forall (i,j) in D with (ref indices) {
    indices2 += (i,j);
}
t2.stop();

writeln("Serial Time: %n".format(t1.elapsed()));
writeln(indices1.size); 
writeln("Parallel Time: %n".format(t2.elapsed()));
writeln(indices2.size); 
```
This issue is descendant from https://github.com/buddha314/numsuch/issues/71

**Comment by bradcray** [comment]
I think that the bottleneck here is that associative domains, like `indices1` and `indices2` are guarded by a lock so that their internal data structures are kept consistent when being accessed by multiple threads in parallel, as in your t2 loop.  Without some sort of parallel safety guard like this, the internal data structures could get corrupted by having multiple tasks read/write them simultaneously.  Unfortunately, because there's not much happening within your loop outside of that locked operation, there isn't much opportunity for parallel speedup.

Something that's supported by sparse domains, but seemingly not by associative ones at present (as far as I can tell with a quick look), is the ability to provide a slew of indices at once rather than doing it one at a time.  This might permit you to create entirely distinct sub-lists of indices in separate tasks and then to take just one lock when adding the bunch of them into the associative domain.  I suspect that this would not take a huge effort to implement.  However, would it improve upon the serial implementation?  In a sense, the associative domain remains a serial bottleneck in that each task will be adding in its indices one by one, so only if the generation of the indices took a lot of time (as it's not in this example) would you expect to see speedup.  In your real program are the indices generated in a way that's nontrivial / more expensive / likely to support parallel speedup.

**Comment by bradcray** [comment]
You might want to evaluate the potential for parallel improvement by comparing a `for` vs. `forall` loop that reads all the indices from the `cursor()` object as you want to in your real code, but then drops them on the floor rather than adding them to a sparse or associative domain.  I.e., if I were doing this for the example above, I'd try timing comparing `for` vs. `forall` versions of the following:

```chapel
for[all] (i,j) in D {
  const x = (i,j);
}
```

where the forall version should show good speedup.  In your case, you'd want to replace the initialization of `x` with the read of `cursor()` and computation of index/indices that you want to add to the domain.  If you see speedup when using the `forall`, that's promising and suggests that the only/main bottleneck is in the domain expansion itself.  If you don't, then there are other bottlenecks / serializations that will need to be addressed, either in CDO or your use of it.

If the only bottleneck is in the domains, then I think the approach of using `bulkAdd()` should help, noting that the example above uses associative domains (which don't currently support `bulkAdd()`) but that your stated target is sparse domains (which do).  There will be a little bit of work to get each task a private list of indices and to do a bulkAdd() of that list, but I think we should be able to work through those issues.

**Comment by mppf** [comment]
I'm not able to fully catch up on this issue right now (I havn't even read all of Brad's comments) but getting associative domain construction to scale is something I've wrestled with in the past: https://chapel-lang.org/CHIUW/2016/Ferguson-slides.pdf  - but note that I think the approach used there is too specific to that 1 case and as a result was never merged. I hope that in the future, we can use a similar approach but implement it more in Chapel module code and have it support other operations (such as gathering key+value pairs and then "reducing" to combine values for the same key). I also think that `bulkAdd` is probably part of a reasonable solution here.

**Comment by buddha314** [comment]
People need to read this deck if only for the introduction of the term "kilo-tweet"

**Comment by Tshimanga** [comment]
@bradcray the forgetful reader runs through a smallish ~1.2m row postgres table in 40s serially and in 0.9s with `forall` on 8 cores. @mppf I just skimmed through the slides. do you have the code for the adjustments you made somewhere on github? I'd like to try adapting it to our needs.

**Comment by mppf** [comment]
@Tshimanga - the code is in fact on github, but it's very very out of date, and also modifies the runtime, and as a result is probably more of a can of worms than I'd recommend you get in to. Even so, it is here: https://github.com/mppf/chapel/tree/operations2 .

**Comment by bradcray** [comment]
@mppf / @Tshimanga : Correct me if I'm wrong, but I think that the issue noted here relates to growing shared memory associative (or maybe sparse) domains and that Michael, your past work was focused on distributed cases?

For the shared-memory case, on the way in this morning, I found myself wondering whether the default implementations of sparse and/or associative domains/arrays should take the approach of chunking the index space into `here.numPU()`s distinct chunks in order to support #cores concurrency to the data structures.  A potential downside is that it could run into the same sort of indexing overhead that we got in the multi-ddata approach to rectangular arrays.

**Comment by mppf** [comment]
> Correct me if I'm wrong, but I think that the issue noted here relates to growing shared memory associative (or maybe sparse) domains and that Michael, your past work was focused on distributed cases?

That's right, but a significant amount of my effort was about increasing the amount of work done each time the lock on the associative domain was taken.

> For the shared-memory case, on the way in this morning, I found myself wondering whether the default implementations of sparse and/or associative domains/arrays should take the approach of chunking the index space into here.numPU()s distinct chunks in order to support #cores concurrency to the data structures. A potential downside is that it could run into the same sort of indexing overhead that we got in the multi-ddata approach to rectangular arrays.

I think that could work, but would rapidly run into the same issues I faced with the distributed version, which arguably already did that.

**Comment by ben-albrecht** [comment]
A next step forward might be to open a PR adding a performance test for this code, so that we can reference the relative performance on a known machine and start gathering data to compare against as improvements are made.

**Comment by marcoscleison** [comment]
Hi all. 
I need to mention that CDO cursor loads all the rows returned from the database in an internal array.
This internal array with the results is split among tasks using a Standard iterator.
@ben-albrecht suggested by chat that we could use zip iterators. But this can be done with leader iterator.
I need to implement it.

**Comment by Tshimanga** [comment]
An update on this issue. `push_back()`ing is still horrendously slow but using array slicing with either zippers or `with (+ reduce j)` brings the parallel performance up to slightly better than serial. Still, there seems to be a lot of potential lost that is worth looking into. Also, using `with (ref j)` results in data leakage; the severity of which depends on the contents of each iteration. @bradcray @mppf  @ben-albrecht @marcoscleison
